PRS and ERS don't promote replicas taking backups

ejortegau opened the pull request:
## Description

This PR changes ERS and PRS so that they prefer not promoting hosts that are currently taking backups. 

The implementation follows what was suggesteed in the RFC of the issue (link below). Namely, the RPCs used to get information about candidates now include an extra field indicating whether they are running backups or not; and that is
used to order the list of promotion candidates.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/16558

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

## Deployment Notes

N/A

frouioui reviewed (commented):
It would be good to include an end-to-end test for this covering several cases: only 1 replica in the cluster, and more than one replica, all replicas being backed up, etc.

Moreover, even though this is not a breaking change per se, we should still document it in the `v22.0` release notes. Which should be put in `./changelog/22.0/22.0.0/summary.md`, the file does not exist yet.

timvaillancourt commented:
@ejortegau I think it would be safer to "prefer" not to promote a `REPLICA` performing backup vs wholly exclude it, the reason is I feel promoting a tablet that is backing-up to `PRIMARY` _(if no other candidates are available)_ is better than doing nothing - which is a 100% outage to writes

If we are down to 1 candidate, still being able to write feels better to me, even if the node is busy with backups as well

GrahamCampbell commented:
Another approach could be to cancel the backup in order to enable promotion, if having backups delayed is preferable to being unavailable for writes for a long time.

GuptaManan100 commented:
I think we should still allow REPLICAs running backups to be promoted, they should just be less preferred. We can do this by changing the sorting logic we use for the tablets in both the promotions (ERS and PRS). If user wants a tablet that is running the backup to not be promoted, they should just change its type to `BACKUP` when they start the backup, then it will never be considered for promotion.

ejortegau commented:
@timvaillancourt & @GuptaManan100 , I have updated the PR so that it works in a "prefer not" way, please have another look. I also added relevant files to Changelog.

timvaillancourt commented:
> @timvaillancourt & @GuptaManan100 , I have updated the PR so that it works in a "prefer not" way, please have another look. I also added relevant files to Changelog. Please have another look.

LGTM! Thanks

ejortegau commented:
@GuptaManan100 's suggestion has been applied, please have another look

deepthi commented on changelog/22.0/22.0.0/summary.md:
```diff
@@ -0,0 +1,17 @@
+
+## <a id="major-changes"/>Major Changes</a>
+
+### <a id="reparents-prefer-not-backing-up"/>Support specifying expected primary in reparents
```
```suggestion
### <a id="reparents-prefer-not-backing-up"/>Avoid promoting a replica that is currently taking a backup
```

deepthi commented on changelog/22.0/22.0.0/summary.md:
```diff
@@ -0,0 +1,17 @@
+
+- **[Known Issues](#known-issues)**
+- **[Major Changes](#major-changes)**
+    - **[Support for specifying expected primary in reparents](#reparents-prefer-not-backing-up)**
```
```suggestion
    - **[Avoid promoting a replica that is currently taking a backup](#reparents-prefer-not-backing-up)**
```

deepthi commented on changelog/22.0/22.0.0/summary.md:
```diff
@@ -0,0 +1,17 @@
+
+### <a id="reparents-prefer-not-backing-up"/>Support specifying expected primary in reparents
+
+The execution of both Planned Shard Reparents and Emergency Reparents now prefer promoting replicas that are not
```
```suggestion
Both planned and emergency failovers now prefer promoting replicas that are not
```

deepthi commented on changelog/22.0/22.0.0/summary.md:
```diff
@@ -0,0 +1,17 @@
+### <a id="reparents-prefer-not-backing-up"/>Support specifying expected primary in reparents
+
+The execution of both Planned Shard Reparents and Emergency Reparents now prefer promoting replicas that are not
+currently taking backups. Notice that if there's only one suitable replica to promote, and it is taking a backup, it
```
```suggestion
currently taking backups. Note that if there's only one suitable replica to promote, and it is taking a backup, it
```

deepthi commented on changelog/22.0/22.0.0/summary.md:
```diff
@@ -0,0 +1,17 @@
+
+The execution of both Planned Shard Reparents and Emergency Reparents now prefer promoting replicas that are not
+currently taking backups. Notice that if there's only one suitable replica to promote, and it is taking a backup, it
+will still be promoted.
```
```suggestion
will still be promoted. This does not apply to `builtin` backups. A replica that is currently taking a `builtin` backup will never be promoted.
```

deepthi commented on changelog/22.0/22.0.0/changelog.md:
```diff
@@ -0,0 +1,5 @@
+# Changelog of Vitess v22.0.0
```
I believe this file will be generated when we do the first release candidate for v22. No need to create it right now.

deepthi commented on proto/tabletmanagerdata.proto:
```diff
@@ -536,6 +537,7 @@ message StopReplicationAndGetStatusResponse {
 
   // Status represents the replication status call right before, and right after telling the replica to stop.
   replicationdata.StopReplicationStatus status = 2;
+  bool backingUp = 3;
```
```suggestion
  bool backup_running = 3;
```

deepthi commented on proto/replicationdata.proto:
```diff
@@ -50,6 +50,7 @@ message Status {
   bool has_replication_filters = 22;
   bool ssl_allowed = 23;
   bool replication_lag_unknown = 24;
+  bool backingUp = 25;
```
```suggestion
  bool backup_running = 25;
```

ejortegau replied:
Fixed.

deepthi commented on proto/replicationdata.proto:
```diff
@@ -65,6 +66,7 @@ message Configuration {
 message StopReplicationStatus {
   replicationdata.Status before = 1;
   replicationdata.Status after = 2;
+  bool backingUp = 3;
```
```suggestion
  bool backup_running = 3;
```

ejortegau replied:
Fixed.

deepthi commented on go/vt/vtctl/reparentutil/emergency_reparenter.go:
```diff
@@ -420,9 +421,11 @@ func (erp *EmergencyReparenter) findMostAdvanced(
 	// We have already removed the tablets with errant GTIDs before calling this function. At this point our winning position must be a
-	// superset of all the other valid positions. If that is not the case, then we have a split brain scenario, and we should cancel the ERS
+	// superset of all the other valid positions - unless the tablet with that position is taking a backup, in which case it might
+	// have a larger GTID set but we are still not choosing it. If that is not the case, then we have a split brain scenario, and
```
```suggestion
	// have a larger GTID set but we are still not choosing it. This can lead to split brain, so
```

deepthi replied:
I had a hard time understanding the previous iteration, I hope this is easier to read.

ejortegau replied:
Fixed.

deepthi commented on go/vt/vtctl/reparentutil/emergency_reparenter.go:
```diff
@@ -420,9 +421,11 @@ func (erp *EmergencyReparenter) findMostAdvanced(
+	// we should cancel the ERS
 	for i, position := range tabletPositions {
-		if !winningPosition.AtLeast(position) {
+		if !winningPosition.AtLeast(position) && !backingUpTablets[topoproto.TabletAliasString(validTablets[i].Alias)] {
```
Idiomatic go would be to do
```
_, ok := backingUpTablets[....]
```
and use `!ok` in the if condition.

ejortegau replied:
This map has all tablets, irrespective of whether they are or are not taking a backup. We need to check the value stored in the map for the alias of the current tablet, so the `if` needs to check what's inside `backingUpTablets[topoproto.TabletAliasString(validTablets[i].Alias)]`. I can add a safety check to ensure that the element is. in the map, though it should always be given how it's created.

ejortegau replied:
Fixed

deepthi commented on go/vt/vtctl/reparentutil/emergency_reparenter_test.go:
```diff
@@ -2859,12 +2860,62 @@ func TestEmergencyReparenter_findMostAdvanced(t *testing.T) {
 				},
 			},
+		}, {
+			name: "choose most advanced not backing up",
```
```suggestion
			name: "choose most advanced not running backup",
```

ejortegau replied:
Fixed.

deepthi commented on go/vt/vtctl/reparentutil/reparent_sorter.go:
```diff
@@ -32,16 +33,18 @@ type reparentSorter struct {
 	tablets          []*topodatapb.Tablet
 	positions        []replication.Position
 	innodbBufferPool []int
+	backingUp        map[string]bool
```
```suggestion
	backupRunning        map[string]bool
```

deepthi replied:
or `runningBackup` 🤷 either one reads better than `backingUp`.

ejortegau replied:
Fixed.

deepthi commented on go/vt/vtctl/reparentutil/reparent_sorter.go:
```diff
@@ -71,6 +74,14 @@ func (rs *reparentSorter) Less(i, j int) bool {
 		return true
 	}
 
+	// We want tablets backing up always at the end of the list, so that's the first thing we check
```
```suggestion
	// We want tablets which are currently running a backup to always be at the end of the list, so that's the first thing we check
```

ejortegau replied:
Fixed.

deepthi commented on go/vt/vtctl/reparentutil/reparent_sorter_test.go:
```diff
@@ -93,54 +94,67 @@ func TestReparentSorter(t *testing.T) {
 		tablets          []*topodatapb.Tablet
 		innodbBufferPool []int
 		positions        []replication.Position
+		backingUpTablets map[string]bool
```
```suggestion
		tabletBackupStatus map[string]bool
```

deepthi replied:
`backingUpTablets` makes me think it's a collection of tablets. This is not a collection of tablets, it's a kind of tablet state.
`tabletBackupState` or `isTabletRunningBackup` or some such thing will also convey the sense of this map properly.

deepthi commented on go/vt/vtctl/reparentutil/util.go:
```diff
@@ -151,17 +153,17 @@ func ElectNewPrimary(
+// findPositionLagBackingUpForTablet processes the replication position and lag for a single tablet and
 // returns it. It is safe to call from multiple goroutines.
-func findPositionAndLagForTablet(ctx context.Context, tablet *topodatapb.Tablet, logger logutil.Logger, tmc tmclient.TabletManagerClient, waitTimeout time.Duration) (replication.Position, time.Duration, error) {
+func findPositionLagBackingUpForTablet(ctx context.Context, tablet *topodatapb.Tablet, logger logutil.Logger, tmc tmclient.TabletManagerClient, waitTimeout time.Duration) (replication.Position, time.Duration, bool, error) {
```
```suggestion
func findPositionLagAndBackupStatusForTablet(ctx context.Context, tablet *topodatapb.Tablet, logger logutil.Logger, tmc tmclient.TabletManagerClient, waitTimeout time.Duration) (replication.Position, time.Duration, bool, error) {
```

deepthi replied:
Maybe we can come up with some shorter function name for this.

ejortegau replied:
I settled for `findTabletPositionLagBackupStatus()`, I hope it's acceptable

deepthi commented on go/vt/vtctl/reparentutil/util_test.go:
```diff
@@ -139,6 +139,115 @@ func TestElectNewPrimary(t *testing.T) {
 			errContains: nil,
 		},
+		{
+			name: "Two good replicas, but one of them backing up so we pick the other one",
```
```suggestion
			name: "Two good replicas, but one of them is running a backup so we pick the other one",
```

deepthi replied:
`Running a backup` / `taking a backup` both sound more idiomatic than `backing up`.

ejortegau replied:
Fixed

deepthi commented on go/vt/vtctl/reparentutil/util_test.go:
```diff
@@ -139,6 +139,115 @@ func TestElectNewPrimary(t *testing.T) {
+			errContains: nil,
+		},
+		{
+			name: "Only one replica, but it's backing up. We still use it.",
```
```suggestion
			name: "Only one replica, but it's running a backup. We still use it.",
```

ejortegau replied:
Fixed

deepthi commented on go/vt/vtctl/reparentutil/util_test.go:
```diff
@@ -139,6 +139,115 @@ func TestElectNewPrimary(t *testing.T) {
+							Cell: "zone1",
+							Uid:  102,
+						},
+						Type: topodatapb.TabletType_REPLICA,
```
@GuptaManan100 I assume we already have tests that verify that tablets of type `BACKUP` are always excluded?

GuptaManan100 replied:
Yes.

deepthi commented on go/vt/vttablet/tabletmanager/rpc_backup.go:
```diff
@@ -205,6 +205,10 @@ func (tm *TabletManager) RestoreFromBackup(ctx context.Context, logger logutil.L
 	return err
 }
 
+func (tm *TabletManager) IsBackingUp() bool {
```
`IsBackupRunning` or `IsRunningBackup` would be better.

deepthi reviewed (commented):
The implementation seems correct. My comments are mostly about naming things etc.

GuptaManan100 commented on go/vt/vtctl/reparentutil/reparent_sorter.go:
```diff
@@ -71,6 +74,14 @@ func (rs *reparentSorter) Less(i, j int) bool {
 	}
 
+	// We want tablets backing up always at the end of the list, so that's the first thing we check
+	if !rs.backingUp[topoproto.TabletAliasString(rs.tablets[i].Alias)] && rs.backingUp[topoproto.TabletAliasString(rs.tablets[j].Alias)] {
+		return true
+	}
+	if rs.backingUp[topoproto.TabletAliasString(rs.tablets[i].Alias)] && !rs.backingUp[topoproto.TabletAliasString(rs.tablets[j].Alias)] {
+		return false
+	}
+
```
This check should be after the position and promotion rule. If a tablet is more advanced it should always come first in the list.

ejortegau replied:
Fixed.

ejortegau commented on changelog/22.0/README.md:
```diff
@@ -0,0 +1,3 @@
+## v22.0
```
Same, autogenerated after running a command indicated in a failing CI check, IIRC.

timvaillancourt commented:
@GuptaManan100 / @ejortegau: I can't see why this would change this behaviour, but I wanted to double check that this ERS behaviour remains true after this change:

> Wait for the primary-elect to catch up to the most advanced replica, if it isn't already the most advanced.

To be specific, if the backing-up `REPLICA` is the most up-to-date in replication _(we've had this happen)_, will the primary-elect catch up to the backing-up `REPLICA` that we likely ignored?

And could any of the backup engines prevent this from happening? For example, some engines take out some heavy mysql-level locks 🤔

GuptaManan100 commented:
The backuping replicas are still viable candidates. We don't remove them from the list even now. It is just an extra piece of information to use to sort the tablets. If we have 2 equally advanced tablets (in terms of position), then the one taking the backup will be lower down the list of candidates. That being said, if a backing up replica is the most advanced tablet there is, we will still promote it.

timvaillancourt commented:
> That being said, if a backing up replica is the most advanced tablet there is, we will still promote it.

@GuptaManan100 at least in our production this would be undesirable. We had this happen on one shard and the overhead of the backup was enough to impact primary operations

I think it would be more ideal for the 2nd-most-current replica to be promoted in this scenario and whatever magic we have to _"catch up to the most advanced replica"_ would be used to catch up to the backing up node. Without this behaviour I think we are open to the same risks that prompted this support. cc @ejortegau for thoughts here

GuptaManan100 commented:
If backup operations impact the tablet from being a primary at all, then we should then decide to never promote it. We can do that by filtering the candidates out in `filterValidCandidates`. We can remove the sorting changes as well. Just removing these tablets would be enough.

If we make this change, then the tablet will be used to get another tablet caught up, but it won't be elected itself. However, with these changes, if we find ourselves in a situation such that the only viable tablet is taking a backup, we will not promote it.

ejortegau commented:
I think this falls back to what behavior we want for the two different cases. Using `filterValidCandates` would work fine for ERS and is in fact what the original version of the PR [used in that case](https://github.com/vitessio/vitess/pull/16997/commits/b2c0b5c3915a88a60925fee526d1472393cd9694#diff-aace4cbbbebad37f5072421e16ca464c0f0b4eb98ad287380c2c764d11e7b836R760-R768). 

This leaves open what should be done for PRS, since PRS logic does not have ideal source selection after picking up the most up to date one as far as I can see. We could do what I proposed [here](https://github.com/vitessio/vitess/pull/16997#issuecomment-2429289798) for that, i.e. fail the PRS with an error that allows the operator to decide whether they will cancel the backup or pick a different source. 

Thoughts?

GuptaManan100 commented:
@ejortegau Yes, we can go back to the original version, I think that would do what we want it to.

As far as PRS is concerned, the primary is online, there is no need for a multi-step promotion. Any candidate that we select, can be promoted. We wait for it to catch up to the primary. So, in PRS, it would be enough to remove these tablets from consideration in `ElectNewPrimary` function.

timvaillancourt commented on go/vt/vtctl/reparentutil/util.go:
```diff
@@ -58,7 +58,8 @@ const (
 // tablet with the most advanced replication position is chosen to minimize the
 // amount of time spent catching up with the current primary. Further ties are
-// broken by the durability rules.
+// broken by the durability rules. Tablets taking backups are excluded from
```
@ejortegau I think this comment is no longer accurate

ejortegau replied:
No, it's not. It reflects the fact that PRS will not promote a tablet that is taking a backup.

ejortegau commented:
I have updated the PR to do the following:
1. In the ERS case, `filterValidCandidates()` does not return hosts taking backups, *unless* there are no better candidates not taking backups. This means ERS can promote hosts taking backups, but only if there are no other candidates that are not.
2. In the PRS case, `ElectNewPrimary()` excludes hosts taking backups. This means PRS can fail if there's only one good candidate and it's taking a backup.

GuptaManan100 reviewed (approved):
Just one comment change, rest looks good to me!

ejortegau commented on go/vt/vtctl/reparentutil/emergency_reparenter.go:
```diff
@@ -420,7 +420,9 @@ func (erp *EmergencyReparenter) findMostAdvanced(
 
 	// We have already removed the tablets with errant GTIDs before calling this function. At this point our winning position must be a
-	// superset of all the other valid positions. If that is not the case, then we have a split brain scenario, and we should cancel the ERS
+	// superset of all the other valid positions - unless the tablet with that position is taking a backup, in which case it might
```
Fixed.

deepthi commented:
Thanks for the contribution and for your patience with the review process.
